### PR TITLE
DM-9996: Final review of SpanSet and Footprint APIs

### PR DESCRIPTION
**Pull request by TallJimbo**
This is an artificial PR that should not be merged; the branches were created simply to make a diff to comment on in GitHub.

### CODE REVIEW

File: src/geom/SpanSet.cc [hunk 1]
```diff
@@ -0,0 +1,828 @@
...
+    auto labeledPair = _makeLabels();
+    // Here we want to check if there is only one label. Since _makeLabels always increments
+    // at the end of each loop, we need to compare against the number 2. I.e. if everything
+    // gets labeled 1, the label counter will increment to 2, and the if statement will always
+    // be false and _makeLabels will fall through to the end and return to here.
+    if (labeledPair.second > 2) {
+        return false;
+    } else {
+        return true;
+    }
```
Comment (TallJimbo) [inline src/geom/SpanSet.cc]: `return labeledPair.second <= 3;` instead of this block?

File: src/geom/SpanSet.cc [hunk 2]
```diff
@@ -0,0 +1,828 @@
...
+                return true;
+            }
+        }
+    }
+    return false;
+}
+
+bool geom::SpanSet::contains(geom::SpanSet const & other) const {
+    // Handle null SpanSet passed as other
+    if (other.size() == 0) {
```
Comment (TallJimbo) [inline src/geom/SpanSet.cc]: `other.empty()`

File: src/geom/SpanSet.cc [hunk 3]
```diff
@@ -0,0 +1,828 @@
...
+    }
+    // Function to check if a SpanSet is entirely contained within this
+    for (auto const & otherSpn : other) {
+        std::size_t counter = 0;
+        for (auto const & spn : _spanVector) {
+            // Check that the end points of the span from other are contained in the
+            // span from this
+            if (spn.contains(geom::Point2I(otherSpn.getMinX(), otherSpn.getY())) &&
+                spn.contains(geom::Point2I(otherSpn.getMaxX(), otherSpn.getY())))
+                ++counter;
```
Comment (TallJimbo) [inline src/geom/SpanSet.cc]: Multi-line `if` statements should use braces, even if the body is only one line.

File: src/geom/SpanSet.cc [hunk 4]
```diff
@@ -0,0 +1,828 @@
...
+        if (spn.contains(point)) {
+            return true;
+        }
+    }
+    return false;
+}
+
+geom::Point2D geom::SpanSet::computeCentroid() const {
+    // Find the centroid of the SpanSet
+    unsigned int n = 0;
```
Comment (TallJimbo) [inline src/geom/SpanSet.cc]: Why `unsigned int` instead of `int` or `std::size_t`?

File: src/geom/SpanSet.cc [hunk 5]
```diff
@@ -0,0 +1,828 @@
...
+             auto mRange = std::equal_range(yRange.first, yRange.second, m, ComparePrimaryRunM());
+             if ((mRange.first == mRange.second)) {
+                 // If a particular m is missing, we known that this y contains
+                 // no good runs; this is equivalent to Kim et al. step 3.2.
+                 goodRuns.clear();
+             } else {
+                 // Consolidate all primary runs at this m so that they don't overlap.
+                 std::list<PrimaryRun> candidateRuns;
+                 int start_x = mRange.first->xmin;
+                 int end_x = mRange.first->xmax;
```
Comment (TallJimbo) [inline src/geom/SpanSet.cc]: Camelcase.

File: src/geom/SpanSet.cc [hunk 6]
```diff
@@ -0,0 +1,828 @@
...
+                 for (auto run = mRange.first+1; run != mRange.second; ++run) {
```
Comment (TallJimbo) [inline src/geom/SpanSet.cc]: Spaces around `+`.

File: src/geom/SpanSet.cc [hunk 7]
```diff
@@ -0,0 +1,828 @@
...
+        return std::make_shared<geom::SpanSet>(this->begin(), this->end());
+    }
+    /* This function must find all the areas in this and not in other. These SpanSets
+     * may be overlapping with this less than other a1|    b1|   a2|    b2|,
+     * with this greater than other b1|    a1|     b2|    a2|,
+     * with this containing other a1|    b1|     b2|     a2|,
+     * or with other containing this  b1|    a1|    a2|   b2|
+     */
+    std::vector<Span> tempVec;
+    int added;
```
Comment (TallJimbo) [inline src/geom/SpanSet.cc]: Could this be `bool`?

File: src/geom/SpanSet.cc [hunk 8]
```diff
@@ -0,0 +1,828 @@
...
+
+    private:
+        SpanSetPersistenceHelper() :
+            spanSetSchema(),
+            spanY(spanSetSchema.addField<int>("y", "The row of the span", "pixel")),
+            spanX0(spanSetSchema.addField<int>("x0", "First column of span (inclusive)", "pixel")),
+            spanX1(spanSetSchema.addField<int>("x1", "Second column of span (inclusive)", "pixel")) {
+                spanSetSchema.getCitizen().markPersistent();
+            }
+    };
```
Comment (TallJimbo) [inline src/geom/SpanSet.cc]: De-indent this class

File: include/lsst/afw/detection/Footprint.h [hunk 9]
```diff
@@ -0,0 +1,383 @@
+
+
```
Comment (TallJimbo) [inline include/lsst/afw/detection/Footprint.h]: Spurious newlines.

File: include/lsst/afw/detection/Footprint.h [hunk 10]
```diff
@@ -0,0 +1,383 @@
...
+     */
+    explicit Footprint(std::shared_ptr<geom::SpanSet> inputSpans,
+                       geom::Box2I const & region=geom::Box2I());
+    explicit Footprint(std::shared_ptr<geom::SpanSet> inputSpans,
+                       afw::table::Schema const & peakSchema,
+                       geom::Box2I const & region=geom::Box2I());
+    explicit Footprint(): lsst::daf::base::Citizen(typeid(this)),
+                          _spans(std::make_shared<geom::SpanSet>()),
+                          _peaks(PeakTable::makeMinimalSchema()),
+			              _region(geom::Box2I()) {}
```
Comment (TallJimbo) [inline include/lsst/afw/detection/Footprint.h]: Add newlines between and docstrings for these.  Weird indentation on the last one.

File: include/lsst/afw/detection/Footprint.h [hunk 11]
```diff
@@ -0,0 +1,383 @@
...
+
+
+    Footprint(Footprint const & other) = default;
+    Footprint(Footprint && ) = default;
```
Comment (TallJimbo) [inline include/lsst/afw/detection/Footprint.h]: Spurious space after `&&`.

File: include/lsst/afw/detection/Footprint.h [hunk 12]
```diff
@@ -0,0 +1,383 @@
...
+
+    virtual ~Footprint() {}
+
+    /** @brief Indicates if this object is a HeavyFootprint
+     */
+    virtual bool isHeavy() const { return false; }
+
+    /** @brief Return a shared pointer to the SpanSet
+      */
+    std::shared_ptr<geom::SpanSet> const & getSpans() const { return _spans;}
```
Comment (TallJimbo) [inline include/lsst/afw/detection/Footprint.h]: Probably better just to return by `std::shared_ptr<geom::SpanSet>` without `const &`.

File: include/lsst/afw/detection/Footprint.h [hunk 13]
```diff
@@ -0,0 +1,383 @@
...
+    void setPeakSchema(afw::table::Schema const & peakSchema) {
+        if (!getPeaks().empty()) {
+            throw LSST_EXCEPT(
+                pex::exceptions::LogicError,
+                "Cannot change the PeakCatalog schema unless it is empty"
+            );
+        }
+        // this syntax doesn't work in Python, which is why this method has to exist
+        getPeaks() = PeakCatalog(peakSchema);
+    }
```
Comment (TallJimbo) [inline include/lsst/afw/detection/Footprint.h]: Move implementation to .cc.

File: include/lsst/afw/detection/Footprint.h [hunk 14]
```diff
@@ -0,0 +1,383 @@
...
+     */
+    void setPeakCatalog(PeakCatalog const & otherPeaks) {
+        if (!getPeaks().empty()) {
+            throw LSST_EXCEPT(
+                pex::exceptions::LogicError,
+                "Cannot change the PeakCatalog unless it is empty"
+            );
+        }
+        getPeaks() = otherPeaks;
+    }
```
Comment (TallJimbo) [inline include/lsst/afw/detection/Footprint.h]: Move implementation to .cc

File: include/lsst/afw/detection/Footprint.h [hunk 15]
```diff
@@ -0,0 +1,383 @@
...
+
+    /**
+     * @brief Return the number of pixels in this Footprint
+     *
+     * This function returns the real number of pixels, not the area of the bbox.
+     */
+    size_t getArea() const { return _spans->getArea(); }
```
Comment (TallJimbo) [inline include/lsst/afw/detection/Footprint.h]: Should use `std::size_t`, I think.

File: include/lsst/afw/detection/Footprint.h [hunk 16]
```diff
@@ -0,0 +1,383 @@
...
+     * @param dy - How much to move in row direction
+     */
+    void shift(int dx, int dy);
+
+    /**
+    * @brief Shift a Footprint by a given extent
+    *
+    * @param d - ExtentI object which gives the dimensions the Footprint should be shifted
+    */
+    void shift(geom::ExtentI const & d) {shift(d.getX(), d.getY());}
```
Comment (TallJimbo) [inline include/lsst/afw/detection/Footprint.h]: Either put spaces after and before braces in all of these one-liners or don't.  (I prefer with spaces; I don't know what clang-format does).

File: include/lsst/afw/detection/Footprint.h [hunk 17]
```diff
@@ -0,0 +1,383 @@
...
+
+    /**
+     * @brief Return the Footprint's bounding box
+     */
+    geom::Box2I getBBox() const { return _spans->getBBox(); }
```
Comment (TallJimbo) [inline include/lsst/afw/detection/Footprint.h]: Why return by value here and const reference below?  I think it's the same issues in play in both cases (I don't have a strong opinion, but we should be consistent).

File: include/lsst/afw/detection/Footprint.h [hunk 18]
```diff
@@ -0,0 +1,383 @@
...
+     *  @param region - Used to set the "region" box of the returned footprint; note that this is
+     *                  NOT the same as the footprint's bounding box.
+     *  @param doClip - If true, clip the new footprint to the region bbox before returning it.
+     */
+    std::shared_ptr<Footprint> transform(
+        std::shared_ptr<image::Wcs> source,
+        std::shared_ptr<image::Wcs> target,
+        geom::Box2I const & region,
+        bool doClip=true
+    ) const;
```
Comment (TallJimbo) [inline include/lsst/afw/detection/Footprint.h]: While we may want this for backwards compatibility, we probably want the `transform` interfaces that `SpanSet` has too.

File: include/lsst/afw/geom/SpanSet.h [hunk 19]
```diff
@@ -0,0 +1,868 @@
...
+
+    /** @brief Perform a set dilation operation, and return a new object
+     *
+     * Dilate a SpanSet with a kernel specified with the stencil parameter
+     *
+     * @param r radius of the stencil, the length is inclusive i.e. 3 ranges from -3 to 3
+     * @param s must be an enumeration of type geom::Stencil. Specifies the shape of the
+                dilation kernel. May be CIRCLE, MANHATTAN, or BOX
+     */
+    std::shared_ptr<SpanSet> dilate(int r, Stencil s = Stencil::CIRCLE) const;
```
Comment (TallJimbo) [inline include/lsst/afw/geom/SpanSet.h]: I'm a little concerned we're calling the `dilate` and `erode` functions the same thing both here and in `Footprint`, where they're in place.  Going by the convention used elsewhere in this file, maybe these should be `eroded` and `dilated`?

File: python/lsst/afw/detection/footprint.cc [hunk 20]
```diff
@@ -0,0 +1,166 @@
...
+ * You should have received a copy of the LSST License Statement and
+ * the GNU General Public License along with this program.  If not,
+ * see <https://www.lsstcorp.org/LegalNotices/>.
+ */
+
+
+#include <cstdint>
+
+#include <pybind11/pybind11.h>
+#include <pybind11/stl.h>
```
Comment (TallJimbo) [inline python/lsst/afw/detection/footprint.cc]: Put pybind11 headers first, and use quotes, not angle brackets.

File: python/lsst/afw/detection/footprint.cc [hunk 21]
```diff
@@ -0,0 +1,166 @@
...
+
+#include "numpy/arrayobject.h"
+#include "ndarray/pybind11.h"
+#include "ndarray/converter.h"
```
Comment (TallJimbo) [inline python/lsst/afw/detection/footprint.cc]: No need to include "ndarray/converter.h"; it's guaranteed to be included by "ndarray/pybind11.h".

File: python/lsst/afw/detection/footprint.cc [hunk 22]
```diff
@@ -0,0 +1,166 @@
...
+#include "lsst/afw/table/io/Persistable.h"
+#include "lsst/afw/table/io/python.h" // for declarePersistableFacade
+#include "lsst/afw/detection/Footprint.h"
+
+namespace py = pybind11;
+using namespace pybind11::literals;
+
+namespace lsst { namespace afw { namespace detection {
+
+namespace {
```
Comment (TallJimbo) [inline python/lsst/afw/detection/footprint.cc]: Add newline after namespace open.

File: python/lsst/afw/detection/footprint.cc [hunk 23]
```diff
@@ -0,0 +1,166 @@
...
+                                  tmpSpan->setMask(*mask, bitmask);
+                              } else {
+                                  foot->getSpans()->setMask(*mask, bitmask);
+                              }
+                          } catch (lsst::pex::exceptions::OutOfRangeError e) {
+                              throw LSST_EXCEPT(lsst::pex::exceptions::OutOfRangeError,
+                                  "Bounds of a Footprint fall outside mask set doClip to force");
+                          }
+                      }
+                      };
```
Comment (TallJimbo) [inline python/lsst/afw/detection/footprint.cc]: Weird lambda indentation.

File: src/detection/Footprint.cc [hunk 24]
```diff
@@ -0,0 +1,411 @@
...
+
+
+#include "lsst/afw/detection/Footprint.h"
+#include "lsst/afw/table/io/CatalogVector.h"
+#include "lsst/afw/table/io/OutputArchive.h"
+
+namespace lsst { namespace afw { namespace detection {
+
+Footprint::Footprint(std::shared_ptr<geom::SpanSet> inputSpans,
+          geom::Box2I const & region): lsst::daf::base::Citizen(typeid(this)),
```
Comment (TallJimbo) [inline src/detection/Footprint.cc]: Weird indentation.

File: src/detection/Footprint.cc [hunk 25]
```diff
@@ -0,0 +1,411 @@
...
+                                       _spans(inputSpans),
+                                       _peaks(PeakTable::makeMinimalSchema()),
+                                       _region(region) {}
+
+Footprint::Footprint(std::shared_ptr<geom::SpanSet> inputSpans,
+          afw::table::Schema const & peakSchema,
+          geom::Box2I const & region): lsst::daf::base::Citizen(typeid(this)),
+                                       _spans(inputSpans),
+                                       _peaks(peakSchema),
+                                       _region(region) {}
```
Comment (TallJimbo) [inline src/detection/Footprint.cc]: More weird indentation.

File: src/detection/Footprint.cc [hunk 26]
```diff
@@ -0,0 +1,411 @@
...
+}
+
+bool Footprint::contains(geom::Point2I const & pix) const {
+    return getSpans()->contains(pix);
+}
+
+std::shared_ptr<Footprint> Footprint::transform(std::shared_ptr<image::Wcs> source,
+                                     std::shared_ptr<image::Wcs> target,
+                                     geom::Box2I const & region,
+                                     bool doClip) const {
```
Comment (TallJimbo) [inline src/detection/Footprint.cc]: Indent to open parent or 4 chars.

File: src/detection/Footprint.cc [hunk 27]
```diff
@@ -0,0 +1,411 @@
...
+    removeOrphanPeaks();
+}
+
+void Footprint::removeOrphanPeaks() {
+    for (auto iter = getPeaks().begin(); iter != getPeaks().end(); ++iter) {
+        if (!getSpans()->contains(geom::Point2I(iter->getIx(), iter->getIy()))) {
+            iter = getPeaks().erase(iter);
+            --iter;
+        }
+    }
```
Comment (TallJimbo) [inline src/detection/Footprint.cc]: Consider using `while` instead to make it easier to avoid the increment.

File: src/geom/SpanSet.cc [hunk 28]
```diff
@@ -0,0 +1,828 @@
...
+            _runNormalize();
+        }
+        _initialize();
+    }
+}
+
+// Construct a SpanSet from a std vector by moving
+geom::SpanSet::SpanSet(std::vector<Span> && vec, bool normalize): _spanVector(std::move(vec)) {
+    // If the incoming vector is zero, should create an empty SpanSet
+    if (_spanVector.size() == 0) {
```
Comment (natelust) [inline src/geom/SpanSet.cc]: I am slightly embarrassed at how poor my c++ was when I started this project for things like this... I had no idea of basic std apis etc

**Comment by jdswinbank** [comment]
This work has been reviewed and this PR is no longer needed.

**Pull request closed without merging**